### PR TITLE
Fix docker user permissions issues

### LINKED ISSUE #1174: System.UnauthorizedAccessException in version 2.42.4 and later
### Version

2.42.4

### Flavor

CLI (Command-Line Interface)

### Platform

Docker / Debian

### Export format

HTML

### Steps to reproduce

1. Install Debian in a VM
2. Install Docker: https://docs.docker.com/engine/install/debian/#install-using-the-repository
3. Run this test script with version 2.42.4:
   ```bash
   token=$(<token.txt)
   id=$(<channel.txt)
   version="tyrrrz/discordchatexporter:$1"
   rm -rf tmp
   mkdir -p tmp
   sudo docker pull "$version"
   sudo docker run -t -v "$(realpath tmp)":/out "$version" export -t "$token" -c "$id" --fuck-russia
   ```
   
Tested and replicated in a fresh Debian VM. I bisected the issue and found that it was introduced in 2.42.4; 2.42.3 was the last working version. The issue is still present in `latest` as of today.

### Details

Expected behavior: Exports the channel.

Actual behavior:

```
2.42.4: Pulling from tyrrrz/discordchatexporter
Digest: sha256:2a4865027ce2f3aed8e8956ade3cee3f6a2ad556b3f9db20f710f871a074fa35
Status: Image is up to date for tyrrrz/discordchatexporter:2.42.4
docker.io/tyrrrz/discordchatexporter:2.42.4
Resolving channel(s)...
Exporting 1 channel(s)...

Server Meta / server-general ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0%

ERROR
System.UnauthorizedAccessException: Access to the path '/out/Nonclassical Theory and Systems - Server Meta - server-general [994207163880833107].html' is denied.
    System.IO.IOException: Permission denied
  at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirError)
  at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String path, OpenFlags flags, Int32 mode, Boolean failForSymlink, Boolean& wasSymlink, Func`4 createOpenException)
  at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, UnixFileMode openPermissions, Int64& fileLength, UnixFileMode& filePermissions, Boolean failForSymlink, Boolean& wasSymlink, Func`4 createOpenException)
  at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
  at System.IO.File.Create(String path)
  at DiscordChatExporter.Core.Exporting.MessageExporter.CreateMessageWriter(String filePath, ExportFormat format, ExportContext context) in /tmp/app/DiscordChatExporter.Core/Exporting/MessageExporter.cs:104
  at DiscordChatExporter.Core.Exporting.MessageExporter.GetWriterAsync(CancellationToken cancellationToken) in /tmp/app/DiscordChatExporter.Core/Exporting/MessageExporter.cs:59
  at DiscordChatExporter.Core.Exporting.MessageExporter.ExportMessageAsync(Message message, CancellationToken cancellationToken) in /tmp/app/DiscordChatExporter.Core/Exporting/MessageExporter.cs:70
  at DiscordChatExporter.Core.Exporting.ChannelExporter.ExportChannelAsync(ExportRequest request, IProgress`1 progress, CancellationToken cancellationToken) in /tmp/app/DiscordChatExporter.Core/Exporting/ChannelExporter.cs:77
  at DiscordChatExporter.Core.Exporting.ChannelExporter.ExportChannelAsync(ExportRequest request, IProgress`1 progress, CancellationToken cancellationToken) in /tmp/app/DiscordChatExporter.Core/Exporting/ChannelExporter.cs:62
  at DiscordChatExporter.Core.Exporting.ChannelExporter.ExportChannelAsync(ExportRequest request, IProgress`1 progress, CancellationToken cancellationToken) in /tmp/app/DiscordChatExporter.Core/Exporting/ChannelExporter.cs:83
  at DiscordChatExporter.Cli.Commands.Base.ExportCommandBase.<>c__DisplayClass63_2.<<ExportAsync>b__2>d.MoveNext() in /tmp/app/DiscordChatExporter.Cli/Commands/Base/ExportCommandBase.cs:227
  at DiscordChatExporter.Cli.Utils.Extensions.ConsoleExtensions.StartTaskAsync(ProgressContext context, String description, Func`2 performOperationAsync) in /tmp/app/DiscordChatExporter.Cli/Utils/Extensions/ConsoleExtensions.cs:53
  at DiscordChatExporter.Cli.Commands.Base.ExportCommandBase.<>c__DisplayClass63_1.<<ExportAsync>b__1>d.MoveNext() in /tmp/app/DiscordChatExporter.Cli/Commands/Base/ExportCommandBase.cs:201
  at System.Threading.Tasks.Parallel.<>c__53`1.<<ForEachAsync>b__53_0>d.MoveNext()
  at DiscordChatExporter.Cli.Commands.Base.ExportCommandBase.<>c__DisplayClass63_0.<<ExportAsync>b__0>d.MoveNext() in /tmp/app/DiscordChatExporter.Cli/Commands/Base/ExportCommandBase.cs:190
  at Spectre.Console.Progress.<>c__DisplayClass27_0.<<StartAsync>b__0>d.MoveNext() in /_/src/Spectre.Console/Live/Progress/Progress.cs:98
  at Spectre.Console.Progress.<>c__DisplayClass28_0`1.<<StartAsync>b__0>d.MoveNext() in /_/src/Spectre.Console/Live/Progress/Progress.cs:133
  at Spectre.Console.Internal.DefaultExclusivityMode.RunAsync[T](Func`1 func) in /_/src/Spectre.Console/Internal/DefaultExclusivityMode.cs:40
  at Spectre.Console.Progress.StartAsync[T](Func`2 action) in /_/src/Spectre.Console/Live/Progress/Progress.cs:116
  at Spectre.Console.Progress.StartAsync(Func`2 action) in /_/src/Spectre.Console/Live/Progress/Progress.cs:96
  at DiscordChatExporter.Cli.Commands.Base.ExportCommandBase.ExportAsync(IConsole console, IReadOnlyList`1 channels) in /tmp/app/DiscordChatExporter.Cli/Commands/Base/ExportCommandBase.cs:180
  at DiscordChatExporter.Cli.Commands.Base.ExportCommandBase.ExportAsync(IConsole console, IReadOnlyList`1 channelIds) in /tmp/app/DiscordChatExporter.Cli/Commands/Base/ExportCommandBase.cs:318
  at DiscordChatExporter.Cli.Commands.ExportChannelsCommand.ExecuteAsync(IConsole console) in /tmp/app/DiscordChatExporter.Cli/Commands/ExportChannelsCommand.cs:25
  at CliFx.CliApplication.RunAsync(ApplicationSchema applicationSchema, CommandInput commandInput) in /_/CliFx/CliApplication.cs:153
  at CliFx.CliApplication.RunAsync(IReadOnlyList`1 commandLineArguments, IReadOnlyDictionary`2 environmentVariables) in /_/CliFx/CliApplication.cs:193
  ```
  
Based on the version in which this bug was introduced, it is most likely unrelated to #978.

### Checklist

- [X] I have looked through existing issues to make sure that this bug has not been reported before
- [X] I have provided a descriptive title for this issue
- [X] I have made sure that that this bug is reproducible on the latest version of the application
- [X] I have provided all the information needed to reproduce this bug as efficiently as possible
- [ ] I have sponsored this project

### PR DESCRIPTION
**Pull request by Silex**
Hello,

Don't merge this yet, it's not ready.

Closes #1174

### CODE REVIEW

**Comment by Silex** [comment]
Ok @Tyrrrz I think you can merge.

I decided to keep the uid:gid of 1000:1000 but make it explicit in the Dockerfile, chown the files so it just works the first time the data folder is created, and provide documentation for those who have different ids.

Please run some tests yourself.

File: DiscordChatExporter.Cli.dockerfile [hunk 1]
```diff
@@ -40,20 +40,19 @@ LABEL org.opencontainers.image.licenses="MIT"
...
+RUN apk add --no-cache icu-libs icu-data-full
 ENV DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=false
 ENV LC_ALL=en_US.UTF-8
 ENV LANG=en_US.UTF-8
 
-# Use a non-root user to ensure that the files shared with the host are accessible by the host user
-# https://github.com/Tyrrrz/DiscordChatExporter/issues/851
-RUN adduser --disabled-password --no-create-home dce
-USER dce
+# Use a non-root user
```
Comment (Tyrrrz) [inline DiscordChatExporter.Cli.dockerfile:48]: Can you please add a comment explaining why `1000` was chosen specifically as the ID?

File: docker-entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env sh
+
+if [ "$(id -u)" = '0' ]; then
+  chown -R dce:dce /out
+  exec su-exec dce "$0" "$@"
+fi
```
Comment (Tyrrrz) [inline docker-entrypoint.sh:6]: Would be nice to have a comment explaining what this block does. I'm not a Linux user, so it's not immediately obvious to me.

**Comment by Murat-U-Saglam** [comment]
Any updates on this? I am getting the same issue when using docker compose within my workflow

File: DiscordChatExporter.Cli.dockerfile (hunk 1 above)
Comment (Silex) [reply DiscordChatExporter.Cli.dockerfile:48]: You chose 1000 when you did `RUN adduser --disabled-password --no-create-home dce` (it's usually the first user id available). And luckily it happens to map with most installations.

I'm just keeping your chosen id and making it explicit.

File: docker-entrypoint.sh (hunk 2 above)
Comment (Silex) [reply docker-entrypoint.sh:6]: Ok I'll add some comments.

File: .docs/Docker.md [hunk 3]
```diff
@@ -61,6 +61,17 @@ For more information, please refer to the [Dockerfile](https://github.com/Tyrrrz
 
 To get your Token and Channel IDs, please refer to [this page](Token-and-IDs.md).
 
+## Permissions issues
+
+This image was designed with a user running as uid:gid of 1000:1000.
+
+If your current user has different ids, and you want to generate files directly editable for your user, you might want to run the container like this:
+
+``` console
```
Comment (Tyrrrz) [inline .docs/Docker.md:70]:
```suggestion
```console
```

File: .docs/Docker.md [hunk 4]
```diff
@@ -61,6 +61,17 @@ For more information, please refer to the [Dockerfile](https://github.com/Tyrrrz
 
 To get your Token and Channel IDs, please refer to [this page](Token-and-IDs.md).
 
+## Permissions issues
+
+This image was designed with a user running as uid:gid of 1000:1000.
+
+If your current user has different ids, and you want to generate files directly editable for your user, you might want to run the container like this:
```
Comment (Tyrrrz) [inline .docs/Docker.md:68]:
```suggestion
If your current user has different IDs, and you want to generate files directly editable for your user, you might want to run the container like this:
```

File: .docs/Docker.md [hunk 5]
```diff
@@ -61,6 +61,17 @@ For more information, please refer to the [Dockerfile](https://github.com/Tyrrrz
 
 To get your Token and Channel IDs, please refer to [this page](Token-and-IDs.md).
 
+## Permissions issues
```
Comment (Tyrrrz) [inline .docs/Docker.md:64]:
```suggestion
## Unix permissions issues
```

File: DiscordChatExporter.Cli.dockerfile (hunk 1 above)
Comment (Tyrrrz) [inline DiscordChatExporter.Cli.dockerfile:48]:
```suggestion
# Use a non-root user to ensure that the files shared with the host are accessible by the host user
# https://github.com/Tyrrrz/DiscordChatExporter/issues/851
# https://github.com/Tyrrrz/DiscordChatExporter/issues/1174
```

File: DiscordChatExporter.Cli.dockerfile [hunk 6]
```diff
@@ -40,20 +40,21 @@ LABEL org.opencontainers.image.licenses="MIT"
...
+RUN addgroup -S -g 1000 dce && adduser -S -H -G dce -u 1000 dce
 
 # This directory is exposed to the user for mounting purposes, so it's important that it always
 # stays the same for backwards compatibility.
 WORKDIR /out
 
 COPY --from=build /tmp/app/DiscordChatExporter.Cli/bin/publish /opt/app
-ENTRYPOINT ["/opt/app/DiscordChatExporter.Cli"]
\ No newline at end of file
+COPY docker-entrypoint.sh /
```
Comment (Tyrrrz) [inline DiscordChatExporter.Cli.dockerfile:59]:
```suggestion
COPY docker-entrypoint.sh /opt/app
```

File: DiscordChatExporter.Cli.dockerfile [hunk 7]
```diff
@@ -40,20 +40,21 @@ LABEL org.opencontainers.image.licenses="MIT"
...
+ENTRYPOINT ["/docker-entrypoint.sh"]
```
Comment (Tyrrrz) [inline DiscordChatExporter.Cli.dockerfile:60]:
```suggestion
ENTRYPOINT ["/opt/app/docker-entrypoint.sh"]
```

**Pull request merged**